docs/CODE_GUIDELINES.md

```diff
@@ -0,0 +1,174 @@
+    codepath, assert the string representation of the CLI command generated by
+    `samCliBuild.ts`
+  - Example 2: instead of mocking various aspects of `sam local invoke`
+    codepath, shift all CONFIG decisions earlier in the codepath and test the
+    real values.
+  - Both examples result in faster, more-readable, more-useful tests, and more
+    test-coverage of real codepaths.
+
+## Code guidelines
+
+- Use maps to group variables, rather than ad-hoc/informal group signals. [Example](https://github.com/aws/aws-toolkit-vscode/blob/abed2c4c7e1329da785190e286e567525afa9da5/src/test/shared/utilities/timeoutUtils.test.ts#L107-L113)
+  - PREFER:
+    ```
+    let topic = {
+      var1 = 'foo',
+      var2 = 'bar',
+    }
+    ```
+  - INSTEAD OF:
+    ```
+    let topicVar1 = 'foo'
+    let topicVar2 = 'bar'
+    ```
+- Where formal grouping is not possible, using a common prefix in names helps
+  discoverability and helps other developers understand "these symbols are
+  related".
+  - PREFER:
+    ```
+    export interface AwsSamDebuggerConfigLoose extends AwsSamDebuggerConfig {
+      ...
+    }
+    ```
+  - INSTEAD OF:
+    ```
+    export interface MorePermissiveAwsSamDebuggerConfig extends AwsSamDebuggerConfig {
+      ...
+    }
+    ```
+- Use module-qualified names, to avoid aliasing members.
+  - PREFER:
+    ```
+    import * as foo from '../foo'
+    const result = foo.Result
+    ```
+  - INSTEAD OF:
+    ```
+    import { Result as FooResult } from '../foo'
+    const result = FooResult
+    ```
+- Use small names for small scopes.
+  - PREFER:
+    ```
+    things.filter(o => o.isFoo)
+    ```
+  - INSTEAD OF:
+    ```
+    things.filter(thing => thing.isFoo)
+    ```
+
+## User settings
+
+- Global scope is shared across all vscode instances *including* remote/SSH
+  instances. Programmatically setting a Global scope config value will raise
+  the onDidChangeConfiguration event for all (local and remote) vscode
+  instances.
+- https://github.com/clangd/vscode-clangd/issues/25#issuecomment-627505249
+  >  We write clangd.path to the user config (local ~/.config/Code/settings.json)
+  >  rather than the machine config (remote ~/.vscode-server/data/Machine/settings.json).
+  > This has two problems:
+  > 1. if you already have a clangd.path in your remote workspace config, then it
+  >    will take precedence over the local user one.
+  > 2. setting clangd.path in your local config, to a path that isn't present
+  >    on the local machine, means that local clangd won't work.
+- Commit introducing machine scope: https://github.com/microsoft/vscode/commit/1fd2993b540ccf11c34234774f7cc53825d60edf
+- Explanation of VSCode settings hierarchy: https://github.com/microsoft/vscode/issues/97616#issuecomment-633671266
+  - package.json scope=machine : https://code.visualstudio.com/api/references/contribution-points#Configuration-property-schema
+- Example: consider the "User" and "Remote" tabs in the VSCode settings UI.
```

Comment:
I'm not sure if it's worth documenting these differences. We don't set all that many values programmatically and as of now, I don't think we set anything outside of `vscode.ConfigurationTarget.Global` level. This seems more geared towards a general VS Code user, not necessarily an extension dev.

Reply:
It is relevant for VSCode Remote feature--and it was very surprising/confusing to discover.

Reply:
We may move some items like this elsewhere. This is more like a "VSCode technical note". But I don't want to hyper-organize it yet.